### PR TITLE
ci: Update actions to avoid Node.js deprecated warning

### PR DESCRIPTION
**Pull request by aleksraiden**
Update all possible actions to latest version, because at spring 2024 all old actions, based on Nodejs 16 will be deprecated. See https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/ 

**Updates**:

- check-typos to [v1.18.2](https://github.com/crate-ci/typos/releases/tag/v1.18.2)
    - Add --sort flag
    - Protect against crash on multi-byte UTF-8 characters when parsed as ASCII
- dorny/paths-filter to [v3](https://github.com/dorny/paths-filter/releases/tag/v3.0.0)
    - Update NodeJS to 20
    - Update bump checkout action to v4 and use setup-node to setup node and cache npm deps 
- actions/setup-go to [v5](https://github.com/actions/setup-go/releases/tag/v5.0.0)
    -  Update NodeJS to 20
-  actions/cache to [v4] (https://github.com/actions/cache/releases/tag/v4.0.0)
    - Update NodeJS
    - save-always flag by @to-s in https://github.com/actions/cache/pull/1242
- docker/login-action to [v3](https://github.com/docker/login-action/releases/tag/v3.0.0)
    - Update NodeJS and all deps
- docker/setup-qemu-action to [v3](https://github.com/docker/setup-qemu-action/releases/tag/v3.0.0)
- docker/setup-buildx-action to [v3](https://github.com/docker/setup-buildx-action/releases/tag/v3.0.0)   
- docker/metadata-action to [v5](https://github.com/docker/metadata-action/releases/tag/v5.5.1)
    - Custom annotations support
    - Allow images input to be empty to output bare tags
    - Dedup and sort labels
 -  docker/build-push-action to [v5](https://github.com/docker/build-push-action/releases/tag/v5.1.0) 
    - Add annotations input
    - Add secret-envs input
    - Support for a minimal [SLSA Provenance](https://slsa.dev/provenance/) attestation
- actions/github-script to [v7](https://github.com/actions/github-script/releases/tag/v7.0.1)
    - Avoid setting baseUrl to undefined when input is not provided
- actions/setup-python to [v5](https://github.com/actions/setup-python/releases/tag/v5.0.0)
- apache/skywalking-eyes to [v5](https://github.com/apache/skywalking-eyes/releases/tag/v0.5.0)
    - add support for AGPL-3.0
    - Upgrade go version to 1.18
    - New Header Template: GPL-3.0-or-later
    - Docker Multiple Architecture Support
 - actions/upload-artifact to [v4](https://github.com/actions/upload-artifact/releases/tag/v4.3.1)
 - actions/setup-go to [v5](https://github.com/actions/setup-go/releases/tag/v5.0.0)

Notes: In some job we use Centos 7 for run actions and this OS cant run Nodejs 20, so I revert in this case actions to old version. In a future, we need to change CentOS for continues work (e.g. before summer 2024)

### CODE REVIEW

File: .github/workflows/kvrocks.yaml [hunk 1]
```diff
@@ -438,11 +440,12 @@ jobs:
           wget https://github.com/Kitware/CMake/releases/download/v3.26.4/cmake-3.26.4-linux-x86_64.sh
           bash cmake-3.26.4-linux-x86_64.sh --skip-license --prefix=/usr
 
-      - uses: actions/checkout@v3
-      - uses: actions/setup-go@v3
+      - uses: actions/checkout@v3 #v4 use Node 20 and not working at CentOS 7
+      - uses: actions/setup-go@v4 #v5 use Node 20 too
```
Comment (tisonkun) [inline .github/workflows/kvrocks.yaml:444]: Let's create an issue for this.

**Review by tisonkun — DISMISSED** [review]
Thank you!

File: .github/workflows/kvrocks.yaml (hunk 1 above)
Comment (aleksraiden) [reply .github/workflows/kvrocks.yaml:444]: Of course, thanks!

**Review by torwig — DISMISSED** [review]

**Comment by tisonkun** [comment]
dorny/paths-filter@v3 is not allowed to be used in apache/kvrocks.

You should open an ASF INFRA ticket to ask for whitelisting this action. I don't have my laptop around so you may search previous PRs to find related comments, or ask for @PragmaTwice @git-hulk help.

**Comment by git-hulk** [comment]
@aleksraiden I have filed a ticket to the ASF team to allow using dorny/paths-filter@v3, refer: https://issues.apache.org/jira/browse/INFRA-25491 

cc @tisonkun

File: .github/workflows/kvrocks.yaml (hunk 1 above)
Comment (aleksraiden) [reply .github/workflows/kvrocks.yaml:444]: Let's talk - https://github.com/apache/kvrocks/issues/2101

**Review by git-hulk — DISMISSED** [review]

**Comment by git-hulk** [comment]
> @aleksraiden I have filed a ticket to the ASF team to allow using dorny/paths-filter@v3, refer: https://issues.apache.org/jira/browse/INFRA-25491
> 
> cc @tisonkun

dorny/paths-filter@v3 is ready to use now.

**Comment by PragmaTwice** [comment]
Try to close and reopen for CI.

**Comment by PragmaTwice** [comment]
> dorny/paths-filter@v3 is ready to use now.

Nope. From the conversation, we need to use `v3.0.0` instead of `v3`.

**Comment by git-hulk** [comment]
> > dorny/paths-filter@v3 is ready to use now.
> 
> Nope. From the conversation, we need to use `v3.0.0` instead of `v3`.

I saw the v3 branch is referred to as v3.0.0, so it should be good to use v3.

https://github.com/dorny/paths-filter/tree/v3

**Comment by PragmaTwice** [comment]
> > > dorny/paths-filter@v3 is ready to use now.
> > 
> > 
> > Nope. From the conversation, we need to use `v3.0.0` instead of `v3`.
> 
> I saw the v3 branch is referred to as v3.0.0, so it should be good to use v3.
> 
> https://github.com/dorny/paths-filter/tree/v3

It is not related to "refer", instead it is just regex matching.

v3 and v3.0.0 is different.

**Comment by PragmaTwice** [comment]
~~Also, you can see the current CI status. It cannot be started.~~

Seems the workflow file is not updated.

**Comment by aleksraiden** [comment]
Try to use dorny/paths-filter v3.0.1

**Comment by tisonkun** [comment]
If I get it correctly, you should use exactly v3.0.0

**Comment by aleksraiden** [comment]
Thanks @tisonkun and @PragmaTwice I fix version to v3.0.0 and all CI jobs are runned OK as I see

**Review by tisonkun — APPROVED** [review]
Thank you!

**Pull request merged**